README.md
# scrape

[![Build Status](https://travis-ci.org/jakewarren/scrape.svg?branch=master)](https://travis-ci.org/jakewarren/scrape/)
[![GitHub release](http://img.shields.io/github/release/jakewarren/scrape.svg?style=flat-square)](https://github.com/jakewarren/scrape/releases])
[![MIT License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](https://github.com/jakewarren/scrape/blob/master/LICENSE)
[![Go Report Card](https://goreportcard.com/badge/github.com/jakewarren/scrape)](https://goreportcard.com/report/github.com/jakewarren/scrape)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=shields)](http://makeapullrequest.com)


A command line scraping utility inspired by [scrape]( https://github.com/jeroenjanssens/data-science-at-the-command-line/blob/master/tools/scrape).

## Features

* Scrape using XPath or CSS selectors
* Process HTML from a URL, STDIN, or a local file
* Extract a particular attribute

## Install
### Option 1: Binary

Download the latest release from [https://github.com/jakewarren/scrape/releases/latest](https://github.com/jakewarren/scrape/releases/latest)

### Option 2: From source

```
go get github.com/jakewarren/scrape
```


## Usage

```
Usage of scrape:
  -A, --agent string   user agent string (default "Mozilla/4.0 (Mozilla/4.0; MSIE 7.0; Windows NT 5.1; SV1; .NET CLR 3.0.04506.30)")
  -a, --attr string    attribute to scrape (default "html")
  -c, --css string     css selector
  -h, --help           usage information
  -k, --insecure       skip SSL verification
  -x, --xpath string   xpath query
```

### Examples:

#### Read from URL:
```
❯ scrape -c "h4 a" -a href "https://www.webscraper.io/test-sites/e-commerce/allinone"
/test-sites/e-commerce/allinone/product/244
/test-sites/e-commerce/allinone/product/269
/test-sites/e-commerce/allinone/product/192
```

#### Read from STDIN:
```
❯ curl -A 'Mozilla/4.0 (Mozilla/4.0; MSIE 7.0; Windows NT 5.1; SV1; .NET CLR 3.0.04506.30)' -s "https://www.webscraper.io/test-sites/e-commerce/allinone" | scrape -x "//h4/a" -a href
/test-sites/e-commerce/allinone/product/223
/test-sites/e-commerce/allinone/product/280
/test-sites/e-commerce/allinone/product/278
```

#### Read from file:
```
❯ scrape -x "//h4/a" /tmp/webscrapetest.html
<a href="/test-sites/e-commerce/allinone/product/223" class="title" title="Aspire E1-510">Aspire E1-510</a>
<a href="/test-sites/e-commerce/allinone/product/280" class="title" title="Lenovo V510 Black">Lenovo V510 Blac...</a>
<a href="/test-sites/e-commerce/allinone/product/278" class="title" title="Lenovo V510 Black">Lenovo V510 Blac...</a>
```




CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.0.2] - 2019-02-18
### Fixed
- Fixed text output when using a CSS selector

## [1.0.1] - 2019-01-17
### Fixed
- Fixed html output when using a CSS selector

## [1.0.0] - 2018-11-06
- Initial Release


main.go
package main

import (
	"bufio"
	"crypto/tls"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"time"

	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
	"github.com/spf13/pflag"
)

// Exit codes
const (
	exitOK = iota
	exitScraping
	exitNoInput
	exitOpenFile
	exitFetchURL
)

// struct to hold config / app data
type config struct {
	xpath     string
	css       string
	userAgent string
	attr      string
	insecure  bool
	in        io.Reader
}

var conf config

func main() {

	helpFlag := pflag.BoolP("help", "h", false, "usage information")
	pflag.StringVarP(&conf.css, "css", "c", "", "css selector")
	pflag.StringVarP(&conf.xpath, "xpath", "x", "", "xpath query")
	pflag.StringVarP(&conf.userAgent, "agent", "A", "Mozilla/4.0 (Mozilla/4.0; MSIE 7.0; Windows NT 5.1; SV1; .NET CLR 3.0.04506.30)", "user agent string")
	pflag.StringVarP(&conf.attr, "attr", "a", "html", "attribute to scrape")
	pflag.BoolVarP(&conf.insecure, "insecure", "k", false, "skip SSL verification")
	pflag.Parse()

	if *helpFlag {
		pflag.Usage()
		os.Exit(exitOK)
	}

	log.Logger = zerolog.New(zerolog.ConsoleWriter{Out: os.Stderr}).With().Timestamp().Logger()
	zerolog.SetGlobalLevel(zerolog.DebugLevel)

	// Determine whether the program's input should be:
	// file, HTTP URL or stdin
	filename := pflag.Arg(0)
	if filename == "" || filename == "-" {

		// no file provided so attempt to read piped data from stdin
		info, err := os.Stdin.Stat()
		if err != nil {
			fatal(exitNoInput, err)
		}
		// ensure stdin is a pipe, bail if not
		if info.Mode()&os.ModeNamedPipe == 0 {
			fatal(exitNoInput, fmt.Errorf("%s", "No input provided to STDIN"))
		}
		conf.in = os.Stdin
	} else {
		if !validURL(filename) { // read from a file
			r, err := os.Open(filename)
			if err != nil {
				fatal(exitOpenFile, err)
			}
			conf.in = r
		} else { // read from URL
			r, err := getURL(filename)
			if err != nil {
				fatal(exitFetchURL, err)
			}
			conf.in = r
		}
	}

	if len(conf.xpath) > 0 { // XPath provided
		err := scrapeXPath()
		if err != nil {
			log.Error().Msgf("%s", err)
			os.Exit(exitScraping)
		}
	} else if len(conf.css) > 0 { // CSS selector provided
		err := scrapeCSS()
		if err != nil {
			log.Error().Msgf("%s", err)
			os.Exit(exitScraping)
		}
	}

}

func fatal(i int, e error) {
	log.Error().Msgf("%s", e)
	os.Exit(i)
}

// validURL performs extremely basic validation of a URL
func validURL(url string) bool {
	r := regexp.MustCompile("(?i)^http(?:s)?://")
	return r.MatchString(url)
}

// getURL takes in a URL and returns an io.Reader
func getURL(url string) (io.Reader, error) {
	tr := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: conf.insecure},
	}

	client := http.Client{
		Transport: tr,
		Timeout:   20 * time.Second,
	}

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("User-Agent", conf.userAgent)

	resp, err := client.Do(req)

	if err != nil {
		return nil, err
	}

	return bufio.NewReader(resp.Body), err
}
